crypto/ecdsa.c

```diff
@@ -1249,3 +1252,56 @@ int ecdsa_sig_from_der(const uint8_t *der, size_t der_len, uint8_t sig[64]) {
 
   return 0;
 }
+
+int ecdsa_get_public_key33(const ecdsa_curve *curve, const uint8_t *priv_key,
+                           uint8_t *pub_key) {
+#ifdef USE_SECP256K1_ZKP_ECDSA
+  if (curve == &secp256k1) {
+    return zkp_ecdsa_get_public_key33(curve, priv_key, pub_key);
+  }
+#endif
+  return tc_ecdsa_get_public_key33(curve, priv_key, pub_key);
+}
+
+int ecdsa_get_public_key65(const ecdsa_curve *curve, const uint8_t *priv_key,
+                           uint8_t *pub_key) {
+#ifdef USE_SECP256K1_ZKP_ECDSA
+  if (curve == &secp256k1) {
+    return zkp_ecdsa_get_public_key65(curve, priv_key, pub_key);
+  }
+#endif
+  return tc_ecdsa_get_public_key65(curve, priv_key, pub_key);
+}
+
+int ecdsa_sign_digest(const ecdsa_curve *curve, const uint8_t *priv_key,
+                      const uint8_t *digest, uint8_t *sig, uint8_t *pby,
+                      int (*is_canonical)(uint8_t by, uint8_t sig[64])) {
+#ifdef USE_SECP256K1_ZKP_ECDSA
+  if (curve == &secp256k1 && is_canonical == NULL) {
+    return zkp_ecdsa_sign_digest(curve, priv_key, digest, sig, pby,
+                                 is_canonical);
+  }
+#endif
+  return tc_ecdsa_sign_digest(curve, priv_key, digest, sig, pby, is_canonical);
+}
+
+int ecdsa_verify_digest(const ecdsa_curve *curve, const uint8_t *pub_key,
+                        const uint8_t *sig, const uint8_t *digest) {
+#ifdef USE_SECP256K1_ZKP_ECDSA
+  if (curve == &secp256k1) {
+    return zkp_ecdsa_verify_digest(curve, pub_key, sig, digest);
+  }
+#endif
+  return tc_ecdsa_verify_digest(curve, pub_key, sig, digest);
+}
+
+int ecdsa_recover_pub_from_sig(const ecdsa_curve *curve, uint8_t *pub_key,
+                               const uint8_t *sig, const uint8_t *digest,
+                               int recid) {
+#ifdef USE_SECP256K1_ZKP_ECDSA
+  if (curve == &secp256k1) {
+    return zkp_ecdsa_recover_pub_from_sig(curve, pub_key, sig, digest, recid);
```

Comment:
So let me get this straight. Based on the described scheme all the other ecdsa functions such as `scalar_multiply()` would be implemented in `native_ecdsa.c`, but their declaration would remain in `ecdsa.h` while there exists a file called `native_ecdsa.h`? That seems confusing.

Reply:
> Based on the described scheme all the other ecdsa functions such as scalar_multiply() would be implemented in native_ecdsa.c, but their declaration would remain in ecdsa.h while there exists a file called native_ecdsa.h?

A would move the declaration of `scalar_multiply()` to `native_ecdsa.h`.

The thing I don't like about the architecture is `ecdsa.c` contains both the ecdsa wrappers and the native implementation of ecdsa. These functions should be in separate files, in my opinion.

Reply:
> Then any code that uses functions that were originally declared in ecdsa.h but do not have a counterpart in the secp256k1 library, i.e. are not wrapped, would have to include native_ecdsa.h instead of ecdsa.h (or both).

Exactly.

> Again seems confusing and technically breaks the API although only a little.

What is confusing about it? Which API does it breaks and how?

I think it's the most natural approach. I see it as `ecdsa.c` being an interface and `zkp_ecdsa.c` and `native_ecdsa.c` being two implementations.

Reply:
> What is confusing about it? Which API does it breaks and how?

It moves certain functions to different header files, such as `scalar_multiply()` and others.

> I think it's the most natural approach. I see it as ecdsa.c being an interface and zkp_ecdsa.c and native_ecdsa.c being two implementations.

If it's meant to be an "interface", then we would also have to add dummy wrappers to `ecdsa.c` for the functions that are in `native_ecdsa.c` only. The caller shouldn't have to deal with the question of whether the implementation is native or non-native.

Reply:
> What is wrong about it? What is confusing about it?

As I stated, the caller shouldn't have to deal with the question of whether the implementation is native or non-native. The whole point of this PR is to simplify the usage of the ecdsa calls.

I am not a fan of creating dummy code where it serves no purpose. So I don't like that option either. This is purely a matter of style. It makes no difference in terms of functionality, usability, maintainability or in any other respect. Feel free to create a separate issue.